statedb/statedb.go

```diff
@@ -1,35 +1,720 @@
+func (v *TreeView) Iterate(callback func(key, value []byte) bool) error {
+	return v.tree.Iterate(nil, callback)
+}
+
+// Root returns the root of the tree, which cryptographically summarises the
+// state of the tree.
+func (v *TreeView) Root() ([]byte, error) {
+	return v.tree.Root(nil)
+}
+
+// Size returns the number of leafs (key-values) that this tree contains.
+func (v *TreeView) Size() (uint64, error) {
+	// NOTE: Tree.Size is currently unimplemented
+	return v.tree.Size(nil), nil
+}
+
+// GenProof generates a proof of existence of the given key for this tree.  The
+// returned values are the leaf value and the proof itself.
+func (v *TreeView) GenProof(key []byte) ([]byte, []byte, error) {
+	return v.tree.GenProof(nil, key)
+}
+
+// Dump exports all the tree leafs in a byte array.
+func (v *TreeView) Dump() ([]byte, error) {
+	return v.tree.Dump()
+}
+
+// subTree is an internal function used to open the subTree (singleton and
+// non-singleton) as a TreeView.  The treeView.db is created from
+// v.db appending the prefix `subKeySubTree | cfg.prefix`.  In turn
+// the treeView.db uses the db.Database from treeView.db appending the
+// prefix `'/' | subKeyTree`.  The treeView.tree is opened as a snapshot from
+// the root found in its parent leaf
+func (v *TreeView) subTree(cfg *treeConfig) (treeView *TreeView, err error) {
+	parentLeaf, err := v.tree.Get(nil, cfg.parentLeafKey)
+	if err != nil {
+		return nil, err
+	}
+	root, err := cfg.parentLeafGetRoot(parentLeaf)
+	if err != nil {
+		return nil, err
+	}
+
+	db := subDB(v.db, join(subKeySubTree, cfg.prefix))
+	tx := db.ReadTx()
+	defer tx.Discard()
+	txTree := subReadTx(tx, subKeyTree)
+	defer txTree.Discard()
+	tree, err := tree.New(&readOnlyWriteTx{txTree},
+		tree.Options{DB: subDB(db, subKeyTree), MaxLevels: cfg.maxLevels, HashFunc: cfg.hashFunc})
+	if err == ErrReadOnly {
+		return nil, ErrEmptyTree
+	} else if err != nil {
+		return nil, err
+	}
+	tree, err = tree.FromRoot(root)
+	if err != nil {
+		return nil, err
+	}
+	return &TreeView{
+		db:   db,
+		tree: tree,
+		cfg:  mainTreeCfg,
+	}, nil
+}
+
+// SubTreeSingle returns a TreeView of a singleton SubTree whose root is stored
+// in the leaf with `cfg.Key()`, and is parametrized by `cfg`.
+func (v *TreeView) SubTreeSingle(c *SubTreeSingleConfig) (*TreeView, error) {
+	return v.subTree(c.treeConfig())
+}
+
+// SubTree returns a TreeView of a non-singleton SubTree whose root is stored
+// in the leaf with `key`, and is parametrized by `cfg`.
+func (v *TreeView) SubTree(key []byte, c *SubTreeConfig) (*TreeView, error) {
+	return v.subTree(c.treeConfig(key))
+}
+
+// TreeUpdate is an opened tree that can be updated.  All updates are stored in
+// an internal transaction (shared with all the opened subTrees) and will only
+// be commited with the TreeTx.Commit function is called.
+type TreeUpdate struct {
+	tx        db.WriteTx
+	txTree    db.WriteTx
+	dirtyTree bool
+	tree      *tree.Tree
+	openSubs  map[string]*TreeUpdate
```

Comment:
I think dirtyTree and openSubs need documenting, as it's not clear why they are needed